test/cadet/updater/xml_parser_test.exs

```diff
@@ -0,0 +1,394 @@
+  alias Cadet.Test.XMLGenerator
+  alias Cadet.Updater.XMLParser
+
+  use Cadet.DataCase
+
+  import Cadet.Factory
+  import ExUnit.CaptureLog
+
+  @local_name "test/fixtures/local_repo"
+  @locations %{mission: "missions", sidequest: "quests", path: "paths", contest: "contests"}
+
+  setup do
+    File.rm_rf!(@local_name)
+
+    on_exit(fn ->
+      File.rm_rf!(@local_name)
+    end)
+
+    assessments =
+      Enum.map(
+        AssessmentType.__enum_map__(),
+        &build(:assessment, type: &1, is_published: true)
+      )
+
+    assessments_with_type = Enum.into(assessments, %{}, &{&1.type, &1})
+
+    questions = build_list(5, :question, assessment: nil)
+
+    %{
+      assessments: assessments,
+      questions: questions,
+      assessments_with_type: assessments_with_type
+    }
+  end
+
+  describe "Pure XML Parser" do
+    test "XML Parser happy path", %{assessments: assessments, questions: questions} do
+      for assessment <- assessments do
+        xml = XMLGenerator.generate_xml_for(assessment, questions)
+
+        assert XMLParser.parse_xml(xml) == :ok
+
+        number = assessment.number
+
+        assessment_db =
+          Assessment
+          |> where(number: ^number)
+          |> Repo.one()
+
+        assert_map_keys(
+          Map.from_struct(assessment),
+          Map.from_struct(assessment_db),
+          ~w(title is_published type summary_short summary_long open_at close_at)a ++
+            ~w(number story reading)a
+        )
+
+        assessment_id = assessment_db.id
+
+        questions_db =
+          Question
+          |> where(assessment_id: ^assessment_id)
+          |> order_by(asc: :display_order)
+          |> Repo.all()
+
+        for {question, question_db} <- Enum.zip(questions, questions_db) do
+          assert_map_keys(
+            Map.from_struct(question_db),
+            Map.from_struct(question),
+            ~w(question type library)a
+          )
+        end
+      end
+    end
+
+    test "happy path existing still closed assessment", %{
+      assessments: assessments,
+      questions: questions
+    } do
+      for assessment <- assessments do
+        still_closed_assessment =
+          Map.from_struct(%{
+            assessment
+            | open_at: Timex.shift(Timex.now(), days: 2),
+              close_at: Timex.shift(Timex.now(), days: 6)
+          })
+
+        %Assessment{}
+        |> Assessment.changeset(still_closed_assessment)
+        |> Repo.insert!()
+
+        xml = XMLGenerator.generate_xml_for(assessment, questions)
+
+        assert XMLParser.parse_xml(xml) == :ok
+      end
+    end
+
+    test "open dates not in ISO8601 DateTime", %{assessments: assessments, questions: questions} do
+      date_strings =
+        Enum.map(
+          ~w({ISO:Basic} {ISOdate} {RFC822} {RFC1123} {ANSIC} {UNIX}),
+          &{&1, Timex.format!(Timex.now(), &1)}
+        )
+
+      for assessment <- assessments,
+          {date_format_string, date_string} <- date_strings do
```

Comment:
Can do this to avoid repeating 2 almost-identical tests:

```
for assessment <- assessments,
    {date_format_string, date_string} <- date_strings,
    time_field <- [:open_at, :close_at] do
  assessment_wrong_date_format = %{assessment | time_field => date_string}
```